api/products.py

```diff
@@ -1,28 +1,308 @@
 
-class Products (metaclass = Singleton):
-    def __init__(self, app):
-        @app.get("/list-products")
-        def get_list_products():
-            return "list-products"
+# Using compiled version of regex for readability and (insignificant)
+# performance boost
+REGEX_MD5 = re.compile(r"^[0-9a-f]{32}$")
 
-        @app.get("/products/count")
-        def get_products_count():
-            return "products/count"
+def validate_post_data(db, form_data, upload_folder):
+    pic_path = None
+    try:
+        prod_name = form_data["prodName"]
+        if len(prod_name) == 0:
+            raise ValueError("Name must be non-empty string")
 
-        @app.get("/products/<int:id>")
-        def get_product_by_id(id):
-            return str(id)
+        prod_price = math.floor(float(form_data["prodPrice"]) * 100)
+        prod_instock = int(form_data["prodInStock"])
 
-        @app.post("/products")
-        def post_product():
-            pass
+        # Not mandatory fields
+        prod_descr = form_data.get("prodDescr", None)
+        pic_md5 = form_data.get("md5", None)
+       
+        if pic_md5 and not REGEX_MD5.search(pic_md5):
+            raise exceptions.BadRequest("Field 'MD5 hash' has invalid value.")
+
+        if prod_price <= 0 or prod_instock < 0:
+            raise exceptions.BadRequest("Numeric fields contain invalid value.")
+
+        return prod_name, prod_descr, prod_price, prod_instock, pic_md5
 
-        @app.delete("/products/all")
-        def delete_products_all():
+    except (KeyError, ValueError, exceptions.BadRequest) as orig_exc:
+        # A required field is not present, cannot be cast into required
+        # type, or supplied file name is invalid
+        try:
+            print("now we will delete the picture")
+            pic_id = db.query("""
+                SELECT pic_id
+                FROM pics
+                WHERE pic_md5 = %s::text
+                LIMIT 1 ;
+            """, (pic_md5,) ).single()
+            print(f"pic_md5 is {pic_md5}")
+            pictures.decrease_picture_count(db, pic_id)
+
+        except Exception as err:
+            # pic_md5 is None or does not correspond to a file in disk
+            # This "except" doesn't do anything, I just thought I should
+            # consider this situation
             pass
 
+        # We raise another exception here, to be caught by main function
+        exceptions.printerr(orig_exc)
+        raise exceptions.BadRequest from orig_exc
+
+def validate_patch_data(form_data):
+    # Without a picture, this should be much easier
+    try:
+        prod_id = form_data["prodId"]
+        prod_name = form_data["prodName"]
+        if len(prod_name) == 0:
+            raise ValueError("Name must be non-empty string")
+
+        prod_price = math.floor(float(form_data["prodPrice"]) * 100)
+        prod_instock = int(form_data["prodInStock"])
+
+        # Not mandatory fields
+        prod_descr = form_data.get("prodDescr", None)
```

Comment:
o get ja retorna None por default
```suggestion
        prod_descr = form_data.get("prodDescr")
```